File: packages/python/dirsql/resolve_config.py
"""Synchronous config resolver for `DirSQL.__dict__`.

Mirrors `DirSQLBuilder::resolve` in the Rust core: explicit kwargs win for
scalars; tables and ignore lists are concatenated; persist is OR-ed;
path-valued config fields resolve relative to the config file's parent.
"""

import os
import tomllib


def resolve_config(
    root, tables, ignore, config, persist, persist_path, extensions=None
):
    """Merge kwargs with a `.dirsql.toml` into the serialized state shape."""
    cfg, cfg_tables, cfg_extensions, cfg_dir = {}, [], [], None
    if config is not None:
        with open(config, "rb") as f:
            doc = tomllib.load(f)
        cfg = doc.get("dirsql") or {}
        cfg_tables = doc.get("table") or []
        cfg_extensions = cfg.get("extension") or []
        cfg_dir = os.path.dirname(os.path.abspath(config))

    def _abs(p):
        # Only reached from the `"<key>" in cfg` branches below, and cfg is
        # non-empty only after a config file was loaded -- which sets cfg_dir.
        base = cfg_dir
        assert base is not None
        return p if os.path.isabs(p) else os.path.join(base, p)

    return {
        "root": root or (_abs(cfg["root"]) if "root" in cfg else cfg_dir),
        "tables": [
            {"ddl": t.ddl, "glob": t.glob, "strict": bool(t.strict)}
            for t in (tables or [])
        ]
        + [
            {
                "ddl": e.get("ddl"),
                "glob": e.get("glob"),
                "strict": bool(e.get("strict")),
            }
            for e in cfg_tables
        ],
        "ignore": list(ignore or []) + list(cfg.get("ignore") or []),
        "persist": bool(persist or cfg.get("persist")),
        "persist_path": persist_path
        or (_abs(cfg["persist_path"]) if "persist_path" in cfg else None),
        # Programmatic extensions first (verbatim paths, mirroring the Rust
        # builder), then config-file `[[dirsql.extension]]` entries with
        # relative paths resolved against the config's parent directory.
        "extensions": [
            {"path": e["path"], "entrypoint": e.get("entrypoint")}
            for e in (extensions or [])
        ]
        + [
            {"path": _abs(e["path"]), "entrypoint": e.get("entrypoint")}
            for e in cfg_extensions
        ],
    }

File: packages/ts/src/resolveConfig.ts
// Synchronous config resolver for `DirSQL.toJSON()`.
//
// Mirrors `DirSQLBuilder::resolve` in the Rust core: explicit options win
// for scalars; tables and ignore lists are concatenated; persist is OR-ed;
// path-valued config fields resolve relative to the config file's parent.

import { readFileSync } from "node:fs";
import { dirname, isAbsolute, resolve as resolvePath } from "node:path";
import { parse as parseToml } from "smol-toml";
import type { DirSQLConfig, DirSQLOptions } from "./index.js";

/**
 * Environment variable the `dirsql interpret` launcher sets, before
 * importing a user's native config module, to the config file's parent
 * directory. A native config that supplies neither `root` nor `config`
 * defaults its scan root to this value -- matching how a `.dirsql.toml`
 * defaults its root to the config's parent directory (#251). Outside the
 * interpret subprocess the variable is unset, so normal SDK use is
 * unaffected.
 */
export const INTERPRET_ROOT_ENV = "DIRSQL_INTERPRET_ROOT";

// biome-ignore lint/suspicious/noExplicitAny: TOML root has a dynamic shape.
type Cfg = Record<string, any>;

export function resolveConfig(options: DirSQLOptions): DirSQLConfig {
  // When no config file is supplied, `cfg` is empty so `cfgDir` is never
  // read (the path-resolution helper is only reached via `cfg.root` /
  // `cfg.persist_path` lookups, both of which short-circuit on an empty cfg).
  let cfg: Cfg = {};
  let cfgTables: Cfg[] = [];
  let cfgDir = "";
  if (options.config !== undefined) {
    const doc = parseToml(readFileSync(options.config, "utf8")) as Cfg;
    cfg = (doc.dirsql ?? {}) as Cfg;
    cfgTables = (Array.isArray(doc.table) ? doc.table : []) as Cfg[];
    cfgDir = dirname(resolvePath(options.config));
  }
  const abs = (p: string) => (isAbsolute(p) ? p : resolvePath(cfgDir, p));

  // Precedence for `root`: explicit option > config-derived (`[dirsql].root`
  // or the config's parent dir) > the interpret launcher's implicit root.
  // The implicit root only applies when neither a root nor a config was
  // given -- i.e. a native config with no `root` (#251). Falls back to ""
  // (the prior behavior) when the var is unset, e.g. normal SDK use.
  const implicitRoot =
    options.config !== undefined
      ? typeof cfg.root === "string"
        ? abs(cfg.root)
        : cfgDir
      : (process.env[INTERPRET_ROOT_ENV] ?? "");

  return {
    root: options.root ?? implicitRoot,
    tables: [
      ...(options.tables ?? []).map((t) => ({
        ddl: t.ddl,
        glob: t.glob,
        strict: t.strict === true,
      })),
      ...cfgTables.map((e) => ({
        ddl: e.ddl as string,
        glob: e.glob as string,
        strict: e.strict === true,
      })),
    ],
    ignore: [...(options.ignore ?? []), ...((cfg.ignore ?? []) as string[])],
    persist: (options.persist ?? false) || cfg.persist === true,
    persistPath:
      options.persistPath ??
      (typeof cfg.persist_path === "string" ? abs(cfg.persist_path) : null),
  };
}
